server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1950,62 +1951,240 @@ public void deletePool(Pool pool) {
             return;
         }
 
-        if (alreadyDeletedPools == null) {
-            alreadyDeletedPools = new HashSet<String>();
-        }
+        log.info("Attempting to delete {} pools...", pools.size());
 
-        if (log.isDebugEnabled()) {
-            log.debug("Delete pools: {}", getPoolIds(pools));
+        // TODO: Remove this and fix the bugs it works around. We absolutely should not be
+        // passing state through the various codepaths like this. It makes things far messier
+        // than they need to be and is resulting in running slow calculations multiple times.
+        if (alreadyDeletedPoolIds == null) {
+            alreadyDeletedPoolIds = new HashSet<String>();
         }
 
-        List<Entitlement> entitlementsToRevoke = new LinkedList<Entitlement>();
-        Set<String> subscriptionIds = new HashSet<String>();
-        Set<Pool> poolsToDelete = new HashSet<Pool>();
+        Set<String> poolIds = new HashSet<String>();
+        Set<String> entitlementIds = new HashSet<String>();
+        Owner owner = null;
 
+        // Convert pools to pool IDs.
+        log.info("Fetching related pools and entitlements...");
         for (Pool pool : pools) {
-            if (log.isDebugEnabled()) {
-                log.debug("Deletion of pool {} will revoke the following entitlements: {}",
-                    pool.getId(), getEntIds(pool.getEntitlements()));
+            if (owner == null) {
+                owner = pool.getOwner();
             }
 
-            // If this is a master pool, we should also be deleting any derived/bonus pools
-            // as well...
-            SourceSubscription srcSub = pool.getSourceSubscription();
-            if (srcSub != null && srcSub.getSubscriptionId() != null &&
-                "master".equals(srcSub.getSubscriptionSubKey())) {
+            poolIds.add(pool.getId());
+        }
+
+        // Fetch pools which are derived from the pools we're going to delete...
+        poolIds.addAll(this.poolCurator.getDerivedPoolIdsForPools(poolIds));
 
-                subscriptionIds.add(srcSub.getSubscriptionId());
+        // Fetch related pools and entitlements (recursively)
+        Collection<String> pids = poolIds;
+        int cachedSize;
+        do {
+            // Fetch entitlement IDs for our set of pools
+            Collection<String> eids = this.poolCurator.getEntitlementIdsForPools(pids);
+
+            // Fetch pools which are derived from these entitlements...
+            pids = this.poolCurator.getPoolIdsForSourceEntitlements(eids);
+
+            // Fetch pools which are derived from the pools we're going to delete...
+            pids.addAll(this.poolCurator.getDerivedPoolIdsForPools(poolIds));
+
+            // Add the new entitlement and pool IDs to our list of things to delete
+            cachedSize = poolIds.size();
+            entitlementIds.addAll(eids);
+            poolIds.addAll(pids);
+        }
+        while (poolIds.size() != cachedSize);
```

Comment:
This is a non-recursive implementation of the logic that would have the deletePools/revokeEntitlements call back into itself. Much of the magic happens in the last three lines of the loop: we cache the size of the poolIds, then add the poolIds from the entitlements to it. If the size changed, we need to make sure we look up more related entitlements. Otherwise, we're done.

Reply:
ok